packages/harmony/src/components/input/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,213 @@
+      ...other
+    } = props
+
+    const characterCount = value !== undefined ? `${value}`.length : 0
+    const nearCharacterLimit = maxLength && characterCount >= 0.9 * maxLength
+
+    // Hide the label when: requested, when the size is small, or when adornment text is present
+    const shouldShowLabel =
+      !hideLabel &&
+      size !== TextInputSize.SMALL &&
+      endAdornmentText === undefined &&
+      startAdornmentText === undefined
+
+    const label = required ? `${labelProp} *` : labelProp
+
+    /**
+     * Since Firefox doesn't support the :has() pseudo selector,
+     * manually track the focused state and use classes for focus, required, and disabled
+     */
+    const [isFocused, handleFocus, handleBlur] = useFocusState(
+      onFocusProp,
+      onBlurProp
+    )
+
+    // Whenever a label isn't visible the placeholder shows in it's place
+    const shouldShowPlaceholder = isFocused || !shouldShowLabel
+
+    const style = {
+      [styles.default]: size === TextInputSize.DEFAULT,
+      [styles.small]: size === TextInputSize.SMALL,
+      [styles.warning]: warningProp || nearCharacterLimit,
+      [styles.error]: error,
+      [styles.focused]: isFocused,
+      [styles.disabled]: disabled,
+      [styles.required]: required
+    }
+
+    const inputRender = (
+      <div className={cn(styles.inputRow, layoutStyles.row)}>
+        <div className={cn(layoutStyles.row, styles.inputContainer)}>
+          {startAdornmentText ? (
+            <Text variant='label' size='l' color='subdued'>
+              {startAdornmentText}
+            </Text>
+          ) : null}
+          <input
+            onFocus={handleFocus}
+            onBlur={handleBlur}
+            ref={ref}
+            className={cn(styles.textInput, inputClassName)}
+            value={value}
+            maxLength={maxLength}
+            disabled={disabled}
+            placeholder={shouldShowPlaceholder ? placeholder : undefined}
+            aria-label={
+              label ?? endAdornmentText ?? startAdornmentText ?? placeholder
+            }
+            {...other}
+          />
+        </div>
+        {endAdornmentText ? (
+          <Text variant='label' size='l' color='subdued'>
+            {endAdornmentText}
+          </Text>
+        ) : null}
+      </div>
+    )
+
+    return (
+      <div className={cn(styles.root, className)}>
+        <div className={cn(styles.inputRoot, inputRootClassName, style)}>
+          {StartIcon ? <StartIcon /> : null}
+          {shouldShowLabel ? (
+            <label className={styles.elevatedLabel}>
+              <Text
+                variant='body'
+                tag='span'
+                className={cn(styles.label, {
+                  [styles.hasValue]: characterCount > 0
+                })}
+              >
+                {label}
+              </Text>
+              {inputRender}
+            </label>
+          ) : (
+            inputRender
+          )}
+
+          {showMaxLength && (
```

Comment:
per our [style guide](https://www.notion.so/audiusproject/Javascript-Typescript-Code-Style-4f34640139fc4a809c3765131bc597ff?pvs=4) which I totally forgot to share with you, use `condition ? <></> : null`